src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -313,6 +319,31 @@ async Task<MsalAccessTokenCacheItem> ITokenCacheInternal.FindAccessTokenAsync(Au
             return null;
         }
 
+        private MsalAccessTokenCacheItem FilterATByKeyId(MsalAccessTokenCacheItem item, AuthenticationRequestParameters authenticationRequest)
+        {
+            if (item == null)
+            {
+                return null;
+            }
+
+            string requestKid = authenticationRequest.AuthenticationScheme.KeyId;
+            if (string.IsNullOrEmpty(item.Kid) && string.IsNullOrEmpty(requestKid))
+            {
+                authenticationRequest.RequestContext.Logger.Verbose("Bearer token found");
+                return item;
+            }
+
+            if (string.Equals(item.Kid, requestKid))
+            {
```

Comment:
how is the key id generated for the pop token? or we can sync offline on this.

Reply:
Sure, let's sync. I haven't coded this part yet. It can be fairly complex, so for now I'll focus on a simple key generation strategy for out .net classic / Windows platform:

- generate a public / private key pair
- store it in a file on disk, encrypted with DPAPI

In the future we need to figure out a strategy for: 

- Linux, Mac (no DPAPI there)
- Confidential Clients (they can't rely on files, since they can be distributed systems)
- Mobile platforms